File: sound.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <stdint.h>

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

int main(int argc, char *argv[]) {
    // Parse command-line arguments
    if (argc < 2) {
        fprintf(stderr, "Usage: %s frequency [duration]\n", argv[0]);
        return 1;
    }

    double freq = atof(argv[1]);
    double duration = argc > 2 ? atof(argv[2]) : 3.0;
    const int sample_rate = 44100;

    // Validate input
    if (freq <= 0 || freq > sample_rate / 2) {
        fprintf(stderr, "Frequency must be between 1 Hz and %d Hz.\n", sample_rate / 2);
        return 1;
    }
    if (duration <= 0) {
        fprintf(stderr, "Duration must be a positive number.\n");
        return 1;
    }

    // Calculate samples and fade parameters
    int num_samples = (int)(duration * sample_rate);
    int fade_samples = (int)(0.01 * sample_rate); // 10ms fade
    const double amplitude = 32767.0; // Max 16-bit amplitude

    // Generate sine wave with fade-in/out
    for (int i = 0; i < num_samples; i++) {
        double t = (double)i / sample_rate;
        double value = sin(2 * M_PI * freq * t);
        
        // Apply fade to avoid clicks
        double gain = 1.0;
        if (i < fade_samples)
            gain = (double)i / fade_samples;
        else if (i >= num_samples - fade_samples)
            gain = (double)(num_samples - i - 1) / fade_samples;

        int16_t sample = (int16_t)(amplitude * value * gain);
        fwrite(&sample, sizeof(int16_t), 1, stdout);
    }

    return 0;
}

File: sound.py
import numpy as np
import sounddevice as sd

def generate_sine_wave(frequency, duration, sample_rate=44100):
    """
    Generate a sine wave.

    :param frequency: Frequency of the wave (Hz)
    :param duration: Duration of the wave (seconds)
    :param sample_rate: Number of samples per second
    :return: Numpy array of the sine wave
    """
    t = np.linspace(0, duration, int(sample_rate * duration), endpoint=False)
    wave = 0.5 * np.sin(2 * np.pi * frequency * t)  # Amplitude is halved for safety
    return wave

def play_sound(wave, sample_rate=44100):
    """
    Play a generated waveform.
    """
    sd.play(wave, samplerate=sample_rate)
    sd.wait()  # Wait until the sound finishes playing

# Parameters
frequency = 440  # A4 note, 440 Hz
duration = 2     # 2 seconds
sample_rate = 44100

# Generate and play the sine wave
sine_wave = generate_sine_wave(frequency, duration, sample_rate)
play_sound(sine_wave)
